File: electron-app-v4/tools/recover-images.py
#!/usr/bin/env python3
"""
recover-images.py — Extract deleted image rows from SQLite free pages.
Usage: python3 tools/recover-images.py [path/to/face_recognition.db]
"""
import os, sys, re, sqlite3, struct

DB_PATH = sys.argv[1] if len(sys.argv) > 1 else \
    os.path.join(os.path.dirname(__file__), '..', '..', 'face_recognition.db')
DB_PATH = os.path.realpath(DB_PATH)

if not os.path.exists(DB_PATH):
    print(f"DB not found: {DB_PATH}")
    sys.exit(1)

# Read raw bytes from DB (and WAL if present)
with open(DB_PATH, 'rb') as f:
    db_bytes = f.read()

wal_path = DB_PATH + '-wal'
wal_bytes = b''
if os.path.exists(wal_path):
    with open(wal_path, 'rb') as f:
        wal_bytes = f.read()
    print(f"WAL: {wal_path}  ({len(wal_bytes)/1024/1024:.1f}MB)")

combined = db_bytes + wal_bytes
print(f"DB: {DB_PATH}  ({len(db_bytes)/1024/1024:.1f}MB)")
print(f"Total scan: {len(combined)/1024/1024:.1f}MB")

# ── Extract printable strings ─────────────────────────────────────────────────
def extract_strings(data, min_len=10):
    results = []
    current = []
    for b in data:
        if 0x20 <= b < 0x7f:
            current.append(chr(b))
        else:
            if len(current) >= min_len:
                results.append(''.join(current))
            current = []
    if len(current) >= min_len:
        results.append(''.join(current))
    return results

strings = extract_strings(combined, min_len=10)
print(f"Extracted {len(strings)} printable strings")

# ── Filter and group image-related strings ────────────────────────────────────
IMG_RE  = re.compile(r'^/[^\x00-\x1f]{5,300}\.(jpg|jpeg|png|webp|heic|heif|gif|bmp)$', re.I)
HASH_RE = re.compile(r'^[0-9a-f]{64}$')
DATE_RE = re.compile(r'^\d{4}-\d{2}-\d{2}[ T]\d{2}:\d{2}:\d{2}')
SCENE_T = {'portrait','indoor','outdoor','group','landscape','event','nature','urban','conference','presentation','other'}

# Collect unique image paths first
all_paths = [s for s in strings if IMG_RE.match(s)]
unique_paths = list(dict.fromkeys(all_paths))  # preserve order, deduplicate
print(f"Distinct image paths: {len(unique_paths)}")

# Try to extract records: for each filepath, scan nearby strings for metadata
records = {}
str_idx = {s: i for i, s in enumerate(strings)}  # first occurrence index

for fp in unique_paths:
    # Skip duplicated UUID paths (the filename duplicate of the filepath)
    basename = os.path.basename(fp)
    rec = {'filepath': fp, 'filename': basename, 'file_hash': None,
           'local_path': None, 'ai_desc': None, 'scene_type': None,
           'taken_at': None}

    idx = str_idx.get(fp, 0)
    window = strings[max(0, idx-5):idx+30]

    for s in window:
        if s == fp: continue
        if not rec['file_hash'] and HASH_RE.match(s):
            rec['file_hash'] = s
        elif not rec['local_path'] and IMG_RE.match(s) and s != fp:
            rec['local_path'] = s
        elif not rec['taken_at'] and DATE_RE.match(s):
            rec['taken_at'] = s[:19]
        elif not rec['scene_type'] and s.lower() in SCENE_T:
            rec['scene_type'] = s.lower()
        elif (not rec['ai_desc'] and len(s) > 40 and ' ' in s
              and s[0].isupper() and not s.startswith('/')):
            rec['ai_desc'] = s

    if fp not in records:
        records[fp] = rec
    else:
        # Merge
        existing = records[fp]
        for k in ('file_hash','local_path','taken_at','scene_type','ai_desc'):
            if rec[k] and not existing[k]:
                existing[k] = rec[k]

print(f"Records assembled: {len(records)}")

# ── Check files exist ──────────────────────────────────────────────────────────
found = sum(1 for r in records.values()
            if os.path.exists(r['filepath']) or os.path.exists(r.get('local_path') or ''))
print(f"Files still on disk: {found}/{len(records)}")

# ── Insert into DB ─────────────────────────────────────────────────────────────
con = sqlite3.connect(DB_PATH)
cur = con.cursor()

cur.execute("SELECT COUNT(*) FROM images")
existing_count = cur.fetchone()[0]
if existing_count > 0:
    print(f"\nimages table already has {existing_count} rows.")
    print("DELETE FROM images first if you want to re-run recovery.")
    con.close()
    sys.exit(0)

inserted = 0
for r in records.values():
    try:
        cur.execute("""
            INSERT OR IGNORE INTO images
              (filepath, filename, file_hash, local_path, visibility,
               ai_description, ai_scene_type, taken_at,
               processed, processed_at, created_at)
            VALUES (?,?,?,?,?, ?,?,?, 1, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
        """, (r['filepath'], r['filename'], r['file_hash'], r['local_path'], 'shared',
              r['ai_desc'], r['scene_type'], r['taken_at']))
        if cur.rowcount:
            inserted += 1
    except Exception as e:
        print(f"  Skip {r['filename']}: {e}")

con.commit()
print(f"\n✓ Re-inserted {inserted} image records.")
print("  Face detection data (bboxes/embeddings) must be re-run via batch processing.")

# Show sample
cur.execute("SELECT id, filename, local_path, ai_description IS NOT NULL FROM images LIMIT 5")
rows = cur.fetchall()
print("\nSample recovered rows:")
for row in rows:
    print(f"  id={row[0]}  {row[1]}  local={bool(row[2])}  has_desc={bool(row[3])}")

con.close()

File: electron-app-v4/tools/recover-images.js
'use strict';

/**
 * recover-images.js — Extract deleted image rows from SQLite free pages
 *
 * After a hard-reset, image rows are deleted but their data remains
 * physically in the database file's free pages. This script:
 *   1. Opens the DB file as raw binary
 *   2. Scans ALL pages (including free-list pages) for recognizable image records
 *   3. Re-inserts found records into the images table
 *
 * Usage: node tools/recover-images.js [path/to/face_recognition.db]
 */

const fs     = require('fs');
const path   = require('path');
const crypto = require('crypto');

const DB_PATH = process.argv[2] ||
  path.join(__dirname, '..', '..', 'face_recognition.db');

if (!fs.existsSync(DB_PATH)) {
  console.error('DB not found:', DB_PATH);
  process.exit(1);
}

// ── Read page size from DB header ─────────────────────────────────────────────
const header = Buffer.alloc(100);
const fd = fs.openSync(DB_PATH, 'r');
fs.readSync(fd, header, 0, 100, 0);
fs.closeSync(fd);

const PAGE_SIZE = header.readUInt16BE(16) || 4096;
const FILE_SIZE = fs.statSync(DB_PATH).size;
const PAGE_COUNT = Math.floor(FILE_SIZE / PAGE_SIZE);
console.log(`DB: ${DB_PATH}`);
console.log(`Page size: ${PAGE_SIZE}  Total pages: ${PAGE_COUNT}  File: ${(FILE_SIZE/1024/1024).toFixed(1)}MB`);

// ── Read all pages ─────────────────────────────────────────────────────────────
const dbBuf = fs.readFileSync(DB_PATH);

// ── Also read WAL file if present ─────────────────────────────────────────────
const WAL_PATH = DB_PATH + '-wal';
let walBuf = null;
if (fs.existsSync(WAL_PATH)) {
  walBuf = fs.readFileSync(WAL_PATH);
  console.log(`WAL: ${WAL_PATH}  (${(walBuf.length/1024/1024).toFixed(1)}MB)`);
}

// ── Extract all strings from raw bytes ────────────────────────────────────────
function extractStrings(buf, minLen = 8) {
  const results = [];
  let start = -1;
  for (let i = 0; i < buf.length; i++) {
    const b = buf[i];
    const printable = (b >= 0x20 && b < 0x7f);
    if (printable) {
      if (start === -1) start = i;
    } else {
      if (start !== -1 && i - start >= minLen) {
        results.push({ offset: start, str: buf.slice(start, i).toString('ascii') });
      }
      start = -1;
    }
  }
  return results;
}

// ── Parse image records from string sequences ─────────────────────────────────
// SQLite stores row data as a sequence of values separated by binary varints.
// The extractable text will appear as adjacent strings in the raw bytes.
// We look for patterns like: <uuid>.jpg\0<uuid>.jpg\0<sha256>\0...\0<local_path>\0

const IMAGE_PATH_RE = /^(\/[^\0]{4,300}\.(jpg|jpeg|png|webp|heic|heif|gif|bmp))$/i;
const HASH_RE = /^[0-9a-f]{64}$/;
const DATE_RE = /^\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}/;

function parseImageRecords(buf, source) {
  const strings = extractStrings(buf, 8);
  const records = [];
  let i = 0;

  while (i < strings.length) {
    const s = strings[i];

    // Look for a server filepath (UUID-style in uploads/)
    if (IMAGE_PATH_RE.test(s.str)) {
      const filepath = s.str;
      let filename   = null;
      let file_hash  = null;
      let local_path = null;
      let ai_desc    = null;
      let scene_type = null;
      let taken_at   = null;

      // Scan forward up to 20 entries for related fields
      for (let j = i + 1; j < Math.min(i + 25, strings.length); j++) {
        const t = strings[j].str;
        if (t === filepath) continue;  // skip duplicate
        if (!filename && IMAGE_PATH_RE.test(t) && path.basename(t) === path.basename(filepath)) {
          filename = path.basename(filepath);
        } else if (!filename && /^[0-9a-f-]{36}\.(jpg|jpeg|png|webp|heic)$/i.test(t)) {
          filename = t;
        } else if (!file_hash && HASH_RE.test(t)) {
          file_hash = t;
        } else if (!local_path && IMAGE_PATH_RE.test(t) && t !== filepath) {
          local_path = t;
        } else if (!taken_at && DATE_RE.test(t)) {
          taken_at = t.slice(0, 19);
        } else if (!ai_desc && t.length > 30 && /[A-Z]/.test(t[0]) && t.includes(' ')) {
          ai_desc = t;
        } else if (!scene_type && ['portrait','indoor','outdoor','group','landscape','event','nature','urban','conference','presentation','other'].includes(t.toLowerCase())) {
          scene_type = t.toLowerCase();
        }
      }

      if (!filename) filename = path.basename(filepath);

      records.push({ filepath, filename, file_hash, local_path, ai_desc, scene_type, taken_at, source });
      i += 2;  // skip past this match
      continue;
    }
    i++;
  }
  return records;
}

// ── Deduplicate by filepath ────────────────────────────────────────────────────
const allRecords = new Map();

const dbRecords  = parseImageRecords(dbBuf, 'db');
const walRecords = walBuf ? parseImageRecords(walBuf, 'wal') : [];

for (const r of [...dbRecords, ...walRecords]) {
  if (!allRecords.has(r.filepath)) {
    allRecords.set(r.filepath, r);
  } else {
    // Merge: keep more complete record
    const existing = allRecords.get(r.filepath);
    if (r.ai_desc && !existing.ai_desc) existing.ai_desc = r.ai_desc;
    if (r.local_path && !existing.local_path) existing.local_path = r.local_path;
    if (r.file_hash && !existing.file_hash) existing.file_hash = r.file_hash;
    if (r.taken_at && !existing.taken_at) existing.taken_at = r.taken_at;
    if (r.scene_type && !existing.scene_type) existing.scene_type = r.scene_type;
  }
}

console.log(`\nFound ${allRecords.size} distinct image records in free pages.`);

if (allRecords.size === 0) {
  console.log('No recoverable records found. Data may have been overwritten.');
  process.exit(0);
}

// ── Check which files still exist ─────────────────────────────────────────────
let existCount = 0, missingCount = 0;
for (const [, r] of allRecords) {
  const fileExists = fs.existsSync(r.filepath) || fs.existsSync(r.local_path || '');
  r.file_exists = fileExists;
  if (fileExists) existCount++; else missingCount++;
}

console.log(`Files on disk: ${existCount} found, ${missingCount} missing`);

// ── Re-insert into DB ──────────────────────────────────────────────────────────
const Database = require('better-sqlite3');
const db = new Database(DB_PATH);

// Verify images table is empty (don't double-insert on second run)
const currentCount = db.prepare('SELECT COUNT(*) AS n FROM images').get().n;
if (currentCount > 0) {
  console.log(`\nimages table already has ${currentCount} rows — aborting to avoid duplicates.`);
  console.log('To force recovery, DELETE FROM images first, then run this script.');
  db.close();
  process.exit(0);
}

const insert = db.prepare(`
  INSERT OR IGNORE INTO images
    (filepath, filename, file_hash, local_path, visibility,
     ai_description, ai_scene_type, taken_at,
     processed, processed_at, created_at)
  VALUES (?,?,?,?,?, ?,?,?, 1, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
`);

let inserted = 0;
const insertAll = db.transaction(() => {
  for (const [, r] of allRecords) {
    try {
      const res = insert.run(
        r.filepath, r.filename, r.file_hash || null, r.local_path || null, 'shared',
        r.ai_desc || null, r.scene_type || null, r.taken_at || null,
      );
      if (res.changes) inserted++;
    } catch (e) {
      console.warn(`  Skip ${r.filename}: ${e.message}`);
    }
  }
});
insertAll();

console.log(`\n✓ Re-inserted ${inserted} image records into images table.`);
console.log(`  NOTE: Face detection data (bboxes, embeddings) is NOT recovered.`);
console.log(`  Re-run batch processing on the images folder to restore face data.`);
console.log(`  AI descriptions that were in the free pages HAVE been preserved.`);

db.close();
